feat(blog): 0.23 release blog post (#2416)
kestra-io/docs: content/blogs/release-0-23.md

</div>

### Unit Test Flows

As workflows grow in complexity, so does the need to test them reliably. Kestra introduces native support for Unit Testing, giving users powerful new capabilities to validate their flows with confidence and precision.
Until now, users relied on manual execution or external scripting to verify behavior. With Unit Test support, you can now define expected outcomes, isolate tasks, and detect regressions early—directly inside your development environment.

Key components of the Unit Testing include:

- **Test Suites**: Create test suites targeting individual tasks or full sections of your flows. Each suite encapsulates a set of test cases, allowing you to cover multiple paths and edge cases without executing the full logic end-to-end.
- **Fixtures**: Define fixtures to simulate inputs or task outputs. This skips real execution for mocked tasks, making your tests faster, deterministic, and easier to isolate.
- **Assertions**: Each test can include assertions to check for expected outcomes—task results, flow variables, execution paths, or any output produced by the flow. This helps ensure your flow behaves correctly under different conditions.

> - **API Access**: You can call the Unit Test programmatically via Kestra API, enabling automation in CI/CD pipelines, custom tooling, or integration with development workflows.

Reviewer: afaik this is not available yet -- @AcevedoR can you confirm? for now, I'll open an issue for 0.24 👍

Reviewer: https://github.com/kestra-io/kestra-ee/issues/3826

Reviewer 2: a simple API is available but not documented

#TODO: add example of unit test with a flow and corresponding test suite
::collapse{title="Unit test example"}

::

<div class="video-container">
  <iframe src="https://www.youtube.com/embed/PLACEHOLDER" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>
</div>

### Outputs in Internal Storage

> Kestra 0.23 introduces a powerful configuration-based capability to move task outputs from in-memory context (by default stored in the database) to Kestra's Internal Storage. This Enterprise Edition feature addresses critical scalability challenges for data-intensive workflows.

Reviewer: ```suggestion
By default, all inputs and outputs processed across all Kestra's workflow executions are stored in the same metadata database. However, some of our Enterprise customers require a higher degree of separation across business units (split by tenants). To address this need for separation, Kestra 0.23 introduces a powerful configuration allowing you to store all execution inputs and outputs in internal storage configured for each tenant. This way, each tenant can store that metadata in separate cloud storage buckets, enabling full data separation across tenants when needed.
```

rewritten as this section is wrong, it's not about scalability but purely about physical data separation across tenants

> By offloading task outputs to Internal Storage, Kestra can maintain execution context for complex workflows with large intermediate data while keeping the database footprint minimal. This configuration-based approach gives administrators fine-grained control over resource utilization and performance optimization.
> 
> This feature is particularly valuable for data-focused workflows that process large volumes of data between tasks, as it removes previous limitations on output size while preserving Kestra's robust execution semantics.

Reviewer: ```suggestion
```

#TODO: Add example of configuration

### Improved dashboard management

This new release enhances the dashboard experience with the ability to configure your own default dashboard - putting you in control of what you see when you first log in. This feature allows you to personalize your monitoring and operational view to match your specific workflow requirements.

With improved dashboard management, you can:
- Set any custom-built dashboard as your default view
- Tailor your monitoring experience to focus on the metrics that matter most to you
- Access your most important information immediately upon login
- Switch between different dashboards based on your current needs

#TODO: screenshots?

### Caching dependencies

[FEATURE 5 DESCRIPTION]

> ### Manage Apps & Dashboard with Git

Reviewer: ```suggestion
### Manage Apps & Dashboards with Git
```

> Kestra 0.23.0 introduces powerful new Git integration tasks that allow you to version control your dashboards and applications, bringing the benefits of Git workflows to your Kestra assets.

Reviewer: ```suggestion
Kestra 0.23.0 introduces a powerful Git Sync, allowing you to version control your Dashboards and Apps code and thus manage all your resources as code.
```

> With the new Git integration for dashboards and apps, you can now:
> 
> - **Version control your dashboards and apps** using `gitSyncDashboard`, `gitPushDashboard`, `gitSyncApps`, and `gitPushApps` tasks
> - **Track changes** to configurations over time
> - **Collaborate** with team members using familiar Git workflows
> - **Roll back** to previous versions when needed

Reviewer: ```suggestion
With the new Git integration for dashboards and apps, you can now:

- **Version control your dashboards and apps** using `git.SyncDashboard`, `git.PushDashboard`, `git.SyncApps`, and `git.PushApps` tasks
- **Track changes** to configurations over time
- **Collaborate** with team members using familiar Git workflows
- **Roll back** to previous versions when needed
```

      WHERE updated_at > CURRENT_DATE - INTERVAL '1 day'
      AND (processed_at IS NULL OR processed_at < updated_at)
    interval: PT5M
    fetchType: FETCH

pluginDefaults:
  - type: io.kestra.plugin.jdbc.postgresql
    values:
      url: "{{ secret('POSTGRES_URL') }}"
      username: "{{ secret('POSTGRES_USERNAME') }}"
      password: "{{ secret('POSTGRES_PASSWORD') }}"
```
::

### HubSpot

We've introduced a comprehensive HubSpot plugin with tasks for managing companies, contacts, and deals. The plugin provides a complete set of operations (Create, Get, Update, Delete, Search) for each entity type, allowing you to seamlessly integrate HubSpot CRM operations into your Kestra workflows with proper authentication and consistent property handling.

::collapse{title="Example of HubSpot integration to query companies"}
```yaml
id: hubspot-query-company
namespace: company.team
tasks:
  - id: search_companies
    type: io.kestra.plugin.hubspot.companies.Search
    apiKey: "{{ secret('HUBSPOT_API_KEY') }}"
    properties:
      - name
      - domain
      - industry
    limit: 10
    sorts:
      - propertyName: "createdate"
        direction: "DESCENDING"
```
::

### Ollama

> We're excited to introduce the new Ollama plugin, which allows you to run Ollama CLI commands directly from your Kestra workflows. This integration makes it easy to automate tasks such as pulling models, running local LLMs, and capturing their outputs as part of your data and AI pipelines.

Reviewer: ```suggestion
We're excited to introduce the new Ollama plugin, which allows you to run Ollama CLI commands directly from your Kestra workflows. This integration can help you pull open-source LLMs into your local environment, interact with them via prompts in your AI pipelines, and shut them down when no longer needed.
```

With the Ollama CLI task, you can:
- Pull and manage models using the Ollama CLI
- Run local LLMs and capture their responses
- Chain Ollama commands with other tasks in your workflow
- Output results to files for downstream processing

::collapse{title="Example using Ollama CLI"}
```yaml
id: ollama_flow
namespace: company.team
tasks:
  - id: ollama_cli
    type: io.kestra.plugin.ollama.cli.OllamaCLI
    commands:
      - ollama pull llama2
      - ollama run llama2 "Tell me a joke about AI" > completion.txt
    outputFiles:
      - completion.txt
```
::

### OpenAI Response

> We've enhanced our OpenAI plugin with a new `Responses` task that integrates OpenAI's latest Responses API – their newest agentic API primitive. This task allows you to create AI-generated responses with built-in tools and structured outputs directly within your workflows.

Reviewer: ```suggestion
We've added a new `Responses` task integrating OpenAI's latest Responses API, allowing you to use tools such as e.g. web search, function calling and structured outputs directly within your AI workflows.
```

> The Responses API combines the simplicity of Chat Completions with powerful agentic capabilities, making it ideal for creating action-oriented applications. The task supports all of OpenAI's built-in tools, including:

Reviewer: ```suggestion
The task supports all of OpenAI's built-in tools, including:
```

- Web search for retrieving real-time information
- File search for analyzing documents

> - Computer use for more complex interactions

Reviewer: ```suggestion
- Structured Outputs
```

You can also format outputs as structured JSON, making it easy to parse and use the generated content in downstream tasks. This is particularly valuable for transforming unstructured requests into structured data that can be directly utilized in your data pipelines.

::collapse{title="Example of OpenAI Responses integration"}

> ```yaml
> id: responses_json_search
> namespace: company.team
> inputs:
>   - id: prompt
>     type: STRING
>     defaults: "List recent trends in workflow orchestration. Return as JSON."
> tasks:
>   - id: trends
>     type: io.kestra.plugin.openai.Responses
>     apiKey: "{{ secret('OPENAI_API_KEY') }}"
>     model: gpt-4.1
>     input: "{{ inputs.prompt }}"
>     text:
>       format:
>         type: json_object
>     toolChoice: required
>     tools:
>       - type: web_search_preview
>   - id: log
>     type: io.kestra.plugin.core.log.Log
>     message: "{{ outputs.trends.outputText }}"
> ```

Reviewer: ```suggestion
```yaml
id: web_search
namespace: company.team

inputs:
  - id: prompt
    type: STRING
    defaults: List recent trends in workflow orchestration

tasks:
  - id: trends
    type: io.kestra.plugin.openai.Responses
    apiKey: "{{ secret('OPENAI_API_KEY') }}"
    model: gpt-4.1-mini
    input: "{{ inputs.prompt }}"
    toolChoice: REQUIRED
    tools:
      - type: web_search_preview

  - id: log
    type: io.kestra.plugin.core.log.Log
    message: "{{ outputs.trends.outputText }}"
```
```

that above example won't work as you can't use web search and structured output at the same time (current limitation of their API)

Reviewer: that above example won't work as you can't use web search and structured output at the same time (current limitation of their API)

::

### Langchain (beta)

- Mention RAG capabilities?

### GitHub Workflow

We're introducing a new GitHub Workflow plugin that allows you to trigger GitHub Actions workflows directly from your Kestra flows.

With the GitHub Workflow plugin, you can:
- Dispatch a GitHub Actions workflow using the `io.kestra.plugin.github.workflows.Dispatch` task
- Pass custom inputs and parameters to your workflow
- Integrate GitHub automation seamlessly with other tasks in your Kestra pipelines

::collapse{title="Example using GitHub Workflow Dispatch"}
```yaml
id: github_workflow_dispatch_flow
namespace: company.team
tasks:
  - id: dispatch_workflow

>     type: io.kestra.plugin.github.workflows.Dispatch

Reviewer: ```suggestion
    type: io.kestra.plugin.github.workflows.RunWorkflow
```

naming TBD, asked via Slack

- **Load** task (`io.kestra.plugin.influxdb.Load`) - Load data points to InfluxDB from an ION file where each record becomes a data point.
- **FluxQuery** task (`io.kestra.plugin.influxdb.FluxQuery`) - Queries InfluxDB using the Flux language, with options to output results as ION internal storage or directly in the execution.
- **InfluxQLQuery** task (`io.kestra.plugin.influxdb.InfluxQLQuery`) - Queries InfluxDB using the InfluxQL language, with the same output options as FluxQuery
- **FluxTrigger** (`io.kestra.plugin.influxdb.FluxTrigger`) - Automatically triggers workflow executions when a Flux query returns results

This integration is particularly useful for IoT data processing, monitoring metrics, and any workflow that involves time series data analysis.

::collapse{title="InfluxDB example"}
```yaml
[EXAMPLE_YAML]
```
::

### GraphQL

We've introduced a new GraphQL plugin that enables integration with GraphQL APIs in your data workflows. The plugin features a `Request` task that allows you to execute GraphQL queries and mutations against any GraphQL endpoint, with full support for authentication headers, variables, and complex queries.

> This plugin is particularly valuable for integrating with modern API-driven services that use GraphQL, allowing you to fetch exactly the data you need without over-fetching or under-fetching. Whether you're connecting to GitHub, Shopify, or any custom GraphQL API, this plugin provides a streamlined way to incorporate that data into your orchestration workflows.

Reviewer: ```suggestion
```
We don't need to explain what GraphQL is, and its possible benefits here

With improved dashboard management, you can:
- Set any custom-built dashboard as your default view
- Tailor your monitoring experience to focus on the metrics that matter most to you
- Access your most important information immediately upon login
- Switch between different dashboards based on your current needs

### Python dependency caching

Kestra 0.23 introduces Python dependency caching, bringing significant improvements to the execution of Python tasks. With this feature, execution times for Python tasks are reduced, as dependencies are cached and reused across runs. You can now use official Python Docker images, and multiple executions of the same task will consistently use the same library versions. There is no need to use virtual environments (venv) for installing requirements, simplifying setup and maintenance.

Under the hood, Kestra uses [uv](https://docs.astral.sh/uv/) for fast dependency resolution and caching. This ensures both speed and compatibility with the Python ecosystem.

> Previously, users needed to install Python dependencies manually using the `beforeCommands` property or using a customer Docker image. For example:

Reviewer 3: ```suggestion
Previously, users needed to install Python dependencies manually using the `beforeCommands` property or using a custom Docker image. For example:
```

Key components of the Unit Testing include:

- **Test Suites**: Create test suites targeting individual flows. Each test suite consists of one or more test cases, allowing you to unit test the same tasks multiple times, e.g., using different flow inputs or different fixtures.
- **Fixtures**: Add fixtures for specific inputs or tasks by mocking specific task states, input values, and namespace files for internal storage files. This allows you to avoid running tasks that might be computationally expensive or not required to run as part of a given test case.
- **Assertions**: Each test case can contain multiple assertions that check if the given task outputs match the expected outputs. There are many assertion operations such as `equalTo`, `notEqualTo`, `greaterThan`, `startsWith`, and more. This helps ensure your flow behaves correctly under different conditions.
- **API Access**: You can call the Unit Test programmatically via Kestra API, enabling automation in CI/CD pipelines, custom tooling, or integration with development workflows.

::collapse{title="Unit Test example"}

> This example demonstrates a simple ETL (Extract, Transform, Load) flow. It sends a Slack notification at the start, downloads order data, extracts and transforms product names to uppercase, and then sends the results back to Slack.

Reviewer 4: I would switch to Anna's server-up example. It's easier to follow especially for a blog. https://github.com/kestra-io/kestra-ee/pull/3881/files

    type: io.kestra.plugin.langchain4j.rag.IngestDocument
    provider:
      type: io.kestra.plugin.langchain4j.provider.GoogleGemini
      modelName: gemini-embedding-exp-03-07
      apiKey: xxx
    embeddings:
      type: io.kestra.plugin.langchain4j.embeddings.KestraKVStore
    drop: true
    fromExternalURLs:
      - https://raw.githubusercontent.com/kestra-io/docs/refs/heads/main/content/blogs/release-0-22.md

  - id: hallucinated_answer
    type: io.kestra.plugin.langchain4j.TextCompletion
    provider:
      type: io.kestra.plugin.langchain4j.provider.GoogleGemini
      modelName: gemini-1.5-flash
      apiKey: xxx
    prompt: Which features were released in Kestra 0.22?

  - id: correct_response_with_rag
    type: io.kestra.plugin.langchain4j.rag.Chat
    chatProvider:
      type: io.kestra.plugin.langchain4j.provider.GoogleGemini
      modelName: gemini-1.5-flash
      apiKey: xxx
    embeddingProvider:
      type: io.kestra.plugin.langchain4j.provider.GoogleGemini
      modelName: gemini-embedding-exp-03-07
      apiKey: xxx
    embeddings:
      type: io.kestra.plugin.langchain4j.embeddings.KestraKVStore
    prompt: Which features were released in Kestra 0.22?
```
::

> ### GitHub Workflow

Reviewer 3: Better for SEO
```suggestion
### GitHub Actions Workflow
```

Author: thanks ! added the new name https://github.com/kestra-io/docs/pull/2416/commits/be7b0e3271061f1f651c82b3a08d7da3d093c4ee

and François already have created the associated PR https://github.com/kestra-io/plugin-github/pull/48

---
title: Kestra 0.23 introduces Multi-Panel Editor, No-Code Redesign, and Unit Testing
description: Kestra 0.23 delivers a multi-panel editor, revamped no-code flow editor, YAML unit testing, new filters, and new plugins for a more productive orchestration experience.
date: 2025-06-17T17:00:00
category: News & Product Updates
author:
  name: Benoit Pimpaud
  image: bpimpaud
image: /blogs/release-0-23.jpg
---

The table below highlights the key features of this release.

| Feature                                   | Description                                                                | Edition |
|-------------------------------------------|----------------------------------------------------------------------------| --- |
| Multi Panel Editor       | Enhanced visual editor with split-screen capabilities for simultaneous code and visual editing | All Edition |

> | No-Code Flow Editor | Create Kestra flows from the new form-based UI without writing code | All Edition |

Reviewer: ```suggestion
| No-Code Forms | Create Kestra flows from the new form-based UI tabs without writing code — included as a dedicated view in the new Multi-Panel Editor | All Edition |
```

Reviewer: @Ben8t I think this seems to be the most natural way to present the feature:
- New split-screen Multi-Panel Flow Editor
- ...which includes No-Code Forms as one of the views that can be opened in multiple new tabs

This means that Multi-Panel editor is the 👑 main feature and No-Code Forms is a new component delivered as part of the Multi-Panel editor. Both worth talking about but the Multi-Panel editor is the main enabler

> | Unit Test Flows       | Built-in testing framework for validating workflow behavior | Enterprise Edition |

Reviewer: ```suggestion
| Unit Tests for Flows       | With Unit Tests, we're introducing a language-agnostic, declarative interface to test your flows with fixtures and assertions, allowing you to run tests directly from the UI and catch regressions before they reach production. | Enterprise Edition |
```
it's not a framework, it's a new resource/feature in Kestra API

I'd also move this to the very top as #1 feature as it's more interesting/new than multi-panel/no-code which people need to see to udnerstand

> | New Filters         | New UI Filters with faster autocompletion and editable as plain text | All Edition |

Reviewer: ```suggestion
| New UI Filters         | UI filters now have a faster autocompletion and are editable as plain text | All Edition |
```

| Outputs in Internal Storage      | Store and manage workflow outputs in isolated tenant storage for complete data separation | Enterprise Edition |
| Improved dashboard management         | Configure your own default dashboard with new customizable KPI charts and adjustable chart widths | All Edition |
| Python Dependency Caching    | Speed up your workflows with automatic caching of script dependencies across executions - just define your dependencies and Kestra handles the rest | All Edition |

> | Manage Apps & Dashboard with Git | Version control your dashboards and apps with Git tasks | All Edition |

Reviewer 5: Apps is EE only

Check the video below for a quick overview of all enhancements.

<div class="video-container">
    <iframe src="[YOUTUBE_VIDEO_URL]" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>
</div>

---

Let's dive into these highlights and other enhancements in more detail.

> ## Feature Highlights

Reviewer: ```suggestion
```

We don't need that H2, everything covered in this blog is a feature highlight

I'd turn all downstream from here as H2 instead to have that included in TOC

With this flexible interface, you can:
- Edit the flow using both Code and form-based editors simultaneously while seeing your changes reflected in real-time in the topology view
- Switch between code and form-based views based on your preference or task complexity
- Track task dependencies through the topology view while making edits
- Reference documentation or blueprints without leaving the Editor

You can customize your experience by opening only the panels you need, creating a workspace tailored to your workflow development style. This enhancement significantly improves productivity by reducing context switching and providing multiple perspectives on your workflows simultaneously.

<div class="video-container">
    <iframe src="https://www.youtube.com/embed/SGlzRmJqFBI" title="Multi Panel Editor" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>
</div>

> ### No-Code Flow Editor
> 
> The No-Code Flow Editor has received a major redesign, making it easier and more intuitive to create and manage your workflows without writing any YAML code. This new version introduces a modern, streamlined interface that improves usability and accelerates flow creation for all users.
> 
> Key improvements include:
> - **New design**: A refreshed, user-friendly layout that simplifies navigation and editing, making it easier to build complex workflows visually.
> - **Fixed issues around arrays and uncovered elements**: The editor now properly handles array fields and previously unsupported or hard-to-edit elements, ensuring you can configure all aspects of your flows directly from the UI.
> - **Enhanced functionality and reliability**: The No-Code Editor has been comprehensively improved with robust validation, intelligent default configurations, and expanded support across all Kestra features. These enhancements deliver increased stability, improved error handling, and a more refined user experience for users at all skill levels.

Reviewer: ```suggestion
## No-Code Forms

The new Multi-Panel editor ships with a significant update to the No-Code Forms. When you open the No-Code view, you can add new tasks, triggers, or flow properties from form-based tabs without writing any YAML code. 

Adding any new task or trigger opens a new No-Code tab, allowing you to edit multiple workflow components at the same time.

Key improvements include:
- **New design**: the new layout simplifies navigation and editing, e.g. adding a task runner configuration to your script task will open a new No-Code tab allowing you to edit the main script and its runtime configuration side-by-side.
- **Improved editing of complex objects**: we've taken great care to ensure that complex objects, such as nested properties and arrays are easy to edit from No-Code forms. 
- **Sensible defaults**: the new No-Code forms make it easy to edit properties that have default values. If you want to revert to a default behavior, the "Clear selection" feature will help you remove your custom overrides.
```

too much fluff, better to include specific examples like e.g. task runner, revert to defaults etc

<div style="position: relative; padding-bottom: calc(48.95833333333333% + 41px); height: 0; width: 100%;"><iframe src="https://demo.arcade.software/99kb4bVvCDnir4V4SxjT?embed&embed_mobile=inline&embed_desktop=inline&show_copy_link=true" title="no_code | Kestra EE" frameborder="0" loading="lazy" webkitallowfullscreen mozallowfullscreen allowfullscreen allow="clipboard-write" style="position: absolute; top: 0; left: 0; width: 100%; height: 100%; color-scheme: light;" ></iframe></div>

> ### Unit Test Flows

Reviewer 5: Isn't Flow Test a beta feature for now?

As workflows grow in complexity, so does the need to test them reliably. Kestra introduces native support for Unit Tests in YAML, allowing you to validate your flows and detect regressions early.
Until now, users could write unit tests in Java, but with the new YAML-based Unit Test support, you can now define expected outcomes, isolate tasks, and detect regressions early—directly inside Kestra using the same YAML format as your flows.

> Key components of the Unit Testing include:

Reviewer: ```suggestion
Key components of a Unit Test:
```

- **Test Suites**: Create test suites targeting individual flows. Each test suite consists of one or more test cases, allowing you to unit test the same tasks multiple times, e.g., using different flow inputs or different fixtures.

> - **Fixtures**: Add fixtures for specific inputs or tasks by mocking specific task states, input values, and namespace files for internal storage files. This allows you to avoid running tasks that might be computationally expensive or not required to run as part of a given test case.

Reviewer: ```suggestion
- **Fixtures**: Add fixtures for specific inputs, tasks or files and avoid running tasks that might be computationally expensive or not required to run as part of a given test case.
```

- **Assertions**: Each test case can contain multiple assertions that check if the given task outputs match the expected outputs. There are many assertion operations such as `equalTo`, `notEqualTo`, `greaterThan`, `startsWith`, and more. This helps ensure your flow behaves correctly under different conditions.
- **API Access**: You can call the Unit Test programmatically via Kestra API, enabling automation in CI/CD pipelines, custom tooling, or integration with development workflows.

::collapse{title="Unit Test example"}

> This example demonstrates a simple ETL (Extract, Transform, Load) flow. It sends a Slack notification at the start, downloads order data, extracts and transforms product names to uppercase, and then sends the results back to Slack.

Reviewer: ```suggestion

```

```yaml
id: etl_toBigQuery
namespace: company.team

tasks:
  - id: send_slack_message_started
    type: io.kestra.plugin.notifications.slack.SlackIncomingWebhook
    url: "https://hooks.slack.com/services/XXXX"
    payload: |
      {
        "text": "{{ flow.namespace }}.{{ flow.id }}: Daily products flow has started"
      }

> 
>   - id: extract
>     type: io.kestra.plugin.core.http.Download
>     uri: https://huggingface.co/datasets/kestra/datasets/raw/main/json/orders.json
> 
>   - id: transform_to_products_name
>     type: io.kestra.plugin.core.debug.Return
>     format: "{{ fromJson(read(outputs.extract.uri)) | jq('.Account.Order[].Product[].\"Product Name\"') }}"
> 
>   - id: transform_to_uppercase
>     type: io.kestra.plugin.core.debug.Return
>     format: "{{ fromJson(outputs.transform_to_products_name.value) | upper }}"
> 
>   - id: send_slack_message_with_data
>     type: io.kestra.plugin.notifications.slack.SlackIncomingWebhook
>     url: "https://hooks.slack.com/services/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"  # _int_test channel https://kestra-io.slack.com/archives/C08LHK1D5EE
>     payload: |
>       {
>         "text": "{{ flow.namespace }}.{{ flow.id }}: Here are all the daily products: {{ outputs.transform_to_uppercase.value }}"
>       }
> ```
> 
> The following test suite demonstrates how we can mock Slack messages and assert the results of the transformations. This allows you to validate the flow's behavior without sending real notifications or fetching live data.
> 
> ```yaml
> id: etl_toBigQuery_extract_testsuite
> namespace: company.team
> flowId: etl_toBigQuery
> testCases:
>   - id: extract_should_return_data
>     type: io.kestra.core.tests.flow.UnitTest
>     fixtures:
>       tasks:
>         - id: send_slack_message_started
>           description: "dont send begin slack message"
>         - id: send_slack_message_with_data
>           description: "dont send end output slack message"
>     assertions:
>       - value: "{{outputs.transform_to_uppercase.value}}"
>         isNotNull: true        
>   - id: extract_should_transform_productNames_to_uppercase_mocked
>     type: io.kestra.core.tests.flow.UnitTest
>     fixtures:
>       tasks:
>         - id: send_slack_message_started
>           description: "dont send begin slack message"
>         - id: send_slack_message_with_data
>           description: "dont send end output slack message"
>         - id: extract
>           description: "dont fetch data from API"
>         - id: transform_to_products_name
>           outputs:
>             value: |
>               [
>                 "my-product-1"
>               ]
>     assertions:
>       - value: "{{outputs.transform_to_uppercase.value}}"
>         contains: "MY-PRODUCT-1"
> ```

Reviewer: ```suggestion
```

Cut that too complex example and replace with simple test to a tutorial flow:

Let’s look at a simple flow checking if a server is up and sending a Slack alert if it’s not:

```yaml
id: microservices-and-apis
namespace: tutorial
description: Microservices and APIs
inputs:
  - id: server_uri
    type: URI
    defaults: https://kestra.io
  - id: slack_webhook_uri
    type: URI
    defaults: https://kestra.io/api/mock
tasks:
  - id: http_request
    type: io.kestra.plugin.core.http.Request
    uri: "{{ inputs.server_uri }}"
    options:
      allowFailed: true
  - id: check_status
    type: io.kestra.plugin.core.flow.If
    condition: "{{ outputs.http_request.code != 200 }}"
    then:
      - id: server_unreachable_alert
        type: io.kestra.plugin.notifications.slack.SlackIncomingWebhook
        url: "{{ inputs.slack_webhook_uri }}"
        payload: |
          {
            "channel": "#alerts",
            "text": "The server {{ inputs.server_uri }} is down!"
          }
    else:
      - id: healthy
        type: io.kestra.plugin.core.log.Log
        message: Everything is fine!
```

Here’s how you might write tests for it:

```yaml
id: test_microservices_and_apis
flowId: microservices-and-apis
namespace: tutorial
testCases:
  - id: server_should_be_reachable
    type: io.kestra.core.tests.flow.UnitTest
    fixtures:
      inputs:
        server_uri: https://kestra.io
    assertions:
      - value: "{{outputs.http_request.code}}"
        equalTo: 200
  - id: server_should_be_unreachable
    type: io.kestra.core.tests.flow.UnitTest
    fixtures:
      inputs:
        server_uri: https://kestra.io/bad-url
      tasks:
        - id: server_unreachable_alert
          description: no Slack message from tests
    assertions:
      - value: "{{outputs.http_request.code}}"
        notEqualTo: 200
```

By default, all flow outputs are stored in the shared metadata database. With this new configuration, you can isolate outputs for each tenant or namespace, making sure that sensitive data is not accessible outside its intended scope.

To enable output storage in Internal Storage for a specific tenant or namespace, add the following to your Kestra configuration file:

```yaml
kestra:
  ee:
    outputs:
      store:
        enabled: true # the default is false
```

If you want to enforce this setting globally for all tenants and namespaces, use the following configuration instead:

```yaml
kestra:
  ee:
    outputs:
      store:
        force-globally: true # the default is false
```

With these options, you can control where flow outputs are stored, improving data governance and compliance for organizations with strict separation requirements.

::alert{type="info"}
Currently, the UI is limited and outputs will not be directly visible if using internal storage. You need to preview them or download them as they are not automatically fetched from the internal storage.
::

> ### Improved Dashboard Management
> 
> This new release enhances the dashboard experience with the ability to configure your own default dashboard - putting you in control of what you see when you first log in. This feature allows you to personalize your monitoring and operational view to match your specific workflow requirements.
> 
> With improved dashboard management, you can:
> - Set any custom-built dashboard as your default view
> - Tailor your monitoring experience to focus on the metrics that matter most to you
> - Access your most important information immediately upon login
> - Switch between different dashboards based on your current needs

Reviewer: ```suggestion
## Customizable Dashboards

This release allows you to personalize your default dashboard with new customizable KPI charts and adjustable chart widths. You can now control what charts and metrics you see when you first log in. 

With improved custom dashboards, you can:
- Set any custom-built dashboard as your default view
- Display only metrics and charts that matter most to you
- Access your most important information immediately upon login
- Switch between different dashboards based on your current needs
```

Under the hood, Kestra uses [uv](https://docs.astral.sh/uv/) for fast dependency resolution and caching. This ensures both speed and compatibility with the Python ecosystem.

Previously, users needed to install Python dependencies manually using the `beforeCommands` property or using a custom Docker image. For example:

```yaml
id: python
namespace: company.team

tasks:
  - id: python
    type: io.kestra.plugin.scripts.python.Script
    containerImage: ghcr.io/kestra-io/pydata:latest
    taskRunner:
      type: io.kestra.plugin.scripts.runner.docker.Docker
    beforeCommands:
      - pip install pandas
    script: |
      from kestra import Kestra
      import pandas as pd
      data = {
        'Name': ['Alice', 'Bob', 'Charlie'],
        'Age': [25, 30, 35]
      }
      df = pd.DataFrame(data)
      print(df)
      print("Average age:", df['Age'].mean())
      Kestra.outputs({"average_age": df['Age'].mean()})
```

> With the new release, you can simply use the `dependencies` property to declare your required libraries. Kestra will handle installation and caching automatically:

Reviewer: ```suggestion
With the new release, you can still use `beforeCommands` as above, but on top of that, you have one more tool at your disposal — the new `dependencies` property, allowing you to declaratively define your required Python packages and let Kestra handle installation and caching automatically:
```

- **Version control your dashboards and apps** using `git.SyncDashboard`, `git.PushDashboard`, `git.SyncApps`, and `git.PushApps` tasks
- **Track configuration changes** over time, managing all your resources as code.
- **Collaborate** with team members using familiar Git workflows
- **Roll back** to previous versions when needed.

::collapse{title="Example of Pulling Apps from Git to Kestra"}

The following flow allows to pull an Apps from a GitHub repository and deploy it to the Kestra instance:

```yaml
id: sync_apps_from_git
namespace: system
tasks:
  - id: git
    type: io.kestra.plugin.ee.git.SyncApps
    delete: true # optional; by default, it's set to false to avoid destructive behavior
    url: https://github.com/kestra-io/apps # required
    branch: main
    username: "{{ secret('GITHUB_USERNAME') }}"
    password: "{{ secret('GITHUB_ACCESS_TOKEN') }}"
triggers:
  - id: every_full_hour
    type: io.kestra.plugin.core.trigger.Schedule
    cron: "0 * * * *"
```
::

## Notable Enhancements

**Ion data format support** with new `IonToParquet` and `IonToAvro` tasks for data conversion, plus `InferAvroSchemaFromIon` for schema generation.

> **Pause Task**: The Pause task now uses a `pauseDuration` property, replacing `delay` and removing `timeout`. When the duration ends, the task proceeds based on the `behavior` property: `RESUME` (default), `WARN`, `FAIL`, or `CANCEL`. Manually resumed tasks always succeed.

Reviewer 5: `timeout` is removed because it exist in the parent task, there is no more specific handling of it but the task could still timeout.

There is also a new `onPause` property to define a task that should run when the flow enter a pause.

> **Pause Task**: The Pause task now uses a `pauseDuration` property, replacing `delay` and removing `timeout`. When the duration ends, the task proceeds based on the `behavior` property: `RESUME` (default), `WARN`, `FAIL`, or `CANCEL`. Manually resumed tasks always succeed.

Reviewer: ```suggestion
**Pause Task**: The Pause task now uses a `pauseDuration` property, replacing `delay` and removing `timeout` because `timeout` is a core property available to all tasks incl. `Pause`. When the `pauseDuration` ends, the task proceeds based on the `behavior` property: `RESUME` (default), `WARN`, `FAIL`, or `CANCEL`. Manually resumed tasks always succeed. Finally, the new `onPause` property allows you to easily define a task that should run whenever the task enters a `PAUSED` state, which is especially useful for sending alerts on paused workflows waiting for approval (i.e. waiting to be manually resumed). 
```

> **Plugin Usage Metrics**: Kestra now provides plugin usage metrics based on execution count. These metrics are compatible with [internal metrics](..docs/09.administrator-guide/03.monitoring) and Prometheus, helping you track plugin usage and performance.
> 
> **Data Backup**: Kestra now supports full execution data backup and restoration, ensuring you can recover all execution-related information for disaster recovery.
> 
> **Account Navigation**: Settings and User Profile are now under Account in the bottom left, below the Tenant switcher. These are user-specific settings.
> 
> **Pebble Function Autocompletion**: When editing Pebble expressions (`{{ ... }}`), function names autocomplete as you type.
> 
> **Worker Information in Task Execution**: Task execution details now show the worker ID, hostname, version, and state. Example: `bbbe25da-06fe-42c2-b50f-4deeba2bb3ba: Hostname=postgres-ee-preview-67c9bbcd56-4fnvr, Version=0.23.0-SNAPSHOT, State=RUNNING`.
> 
> **Secret Filtering**: For Google Cloud Secret Manager, Azure Key Vault, and AWS Secrets Manager, the new `filterOnTags` property lets you filter secrets by tags and sync only those that match.

Reviewer: ```suggestion
**Plugin Usage Metrics**: Kestra now provides plugin usage metrics based on an execution count. These metrics are compatible with [internal metrics](..docs/09.administrator-guide/03.monitoring) and Prometheus, helping you track how plugins are used in your organization.

**Data Backup**: We now support full Backup & Restore, including backup of executions and logs data, ensuring you can recover all execution-related information for disaster recovery.

**Account Navigation**: Settings and User Profile are now located under the Account settings in the bottom left corner, just below the Tenant switcher. 

**Pebble Function Autocompletion**: When editing Pebble expressions (`{{ ... }}`), function names autocomplete as you type.

![image](https://github.com/user-attachments/assets/baa5d52a-a5b6-48f4-9177-de76f20f4f24)

**Worker Information in Task Execution**: Task execution details now show the worker ID, hostname, version, and state. Example: `bbbe25da-06fe-42c2-b50f-4deeba2bb3ba: Hostname=postgres-ee-preview-67c9bbcd56-4fnvr, Version=0.23.0-SNAPSHOT, State=RUNNING`.

**Secret Filtering**: For Google Cloud Secret Manager, Azure Key Vault, and AWS Secrets Manager, the new `filterOnTags` property lets you filter secrets by tags and sync only those that match.
```

Let's add a screenshot for that function autocomplete:

![image](https://github.com/user-attachments/assets/a16cb3ed-4f15-46cb-a98c-b6e88f6dc31d)

The plugins support multiple providers, such as OpenAI, Google Gemini, and others, giving you flexibility to select the best model for your use case.

::collapse{title="Example using Langchain RAG capabilities"}
```yaml
id: rag_demo
namespace: company.team

tasks:
  - id: ingest
    type: io.kestra.plugin.langchain4j.rag.IngestDocument
    provider:
      type: io.kestra.plugin.langchain4j.provider.GoogleGemini
      modelName: gemini-embedding-exp-03-07
      apiKey: xxx
    embeddings:
      type: io.kestra.plugin.langchain4j.embeddings.KestraKVStore
    drop: true
    fromExternalURLs:
      - https://raw.githubusercontent.com/kestra-io/docs/refs/heads/main/content/blogs/release-0-22.md

  - id: hallucinated_answer
    type: io.kestra.plugin.langchain4j.TextCompletion
    provider:
      type: io.kestra.plugin.langchain4j.provider.GoogleGemini
      modelName: gemini-1.5-flash
      apiKey: xxx
    prompt: Which features were released in Kestra 0.22?

  - id: correct_response_with_rag

>     type: io.kestra.plugin.langchain4j.rag.Chat

Reviewer 5: ```suggestion
    type: io.kestra.plugin.langchain4j.rag.ChatCompletion
```

    repository: owner/repository
    workflowId: your_workflow_id
    ref: your_branch_or_tag_name
    inputs:
      foo:bar
```
::

### Jenkins

We're introducing a new Jenkins plugin that enables seamless integration with Jenkins CI/CD pipelines directly from your Kestra workflows. This integration is ideal for teams looking to unify their CI/CD automation and workflow orchestration, enabling end-to-end automation from code to deployment.

With the Jenkins plugin, you can:
- Trigger a Jenkins job build using the `io.kestra.plugin.jenkins.JobBuild` task
- Retrieve detailed information about a Jenkins job with the `io.kestra.plugin.jenkins.JobInfo` task

::collapse{title="Example using Jenkins JobBuild"}
```yaml
id: jenkins_job_trigger
namespace: company.team
tasks:
  - id: build
    type: io.kestra.plugin.jenkins.JobBuild
    jobName: deploy-app
    serverUri: http://localhost:8080
    username: admin
    api_token: "{{ secret('API_TOKEN') }}"
    parameters:
      branch: main
      environment:
        - staging
```
::

### Go Scripts

Kestra 0.23 introduces powerful new capabilities for running Go code with the addition of two dedicated Go script tasks:

> - `Script` task (io.kestra.plugin.scripts.go.Script) - Executes multi-line Go scripts
> - `Commands` task (io.kestra.plugin.scripts.go.Commands) - Runs multiple Go scripts using `go run`

Reviewer: ```suggestion
- `Script` task (io.kestra.plugin.scripts.go.Script) - for inline code
- `Commands` task (io.kestra.plugin.scripts.go.Commands) - for code stored in Namespace Files or passed from a local directory (e.g. cloned from a Git repository) which can be executed using the `go run` command.
```
